cpp/include/ucxx/memory_handle.h

```diff
@@ -0,0 +1,159 @@
+   *
+   * This is the internal implementation of `ucxx::MemoryHandle` constructor, made private
+   * not to be called directly. This constructor is made private to ensure all UCXX objects
+   * are shared pointers and the correct lifetime management of each one.
+   *
+   * Instead the user should use one of the following:
+   *
+   * - `ucxx::Context::createMemoryHandle`
+   * - `ucxx::createMemoryHandle()`
+   *
+   * @throws ucxx::Error if either `ucp_mem_map` or `ucp_mem_query` fail.
+   *
+   * @param[in] context parent context where to map memory.
+   * @param[in] size    the minimum size of the memory allocation
+   * @param[in] buffer  the pointer to an existing allocation or `nullptr` to allocate a
+   *                    new memory region.
+   */
+  MemoryHandle(std::shared_ptr<Context> context, const size_t size, void* buffer);
+
+ public:
+  MemoryHandle()                               = delete;
+  MemoryHandle(const MemoryHandle&)            = delete;
+  MemoryHandle& operator=(MemoryHandle const&) = delete;
+  MemoryHandle(MemoryHandle&& o)               = delete;
+  MemoryHandle& operator=(MemoryHandle&& o)    = delete;
+
+  /**
+   * @brief Constructor for `shared_ptr<ucxx::MemoryHandle>`.
+   *
+   * The constructor for a `shared_ptr<ucxx::MemoryHandle>` object, mapping a memory buffer
+   * with UCP to provide RMA (Remote Memory Access) to.
+   *
+   * The allocation requires a `size` and a `buffer`. The actual size of the allocation may
+   * be larger than requested, and can later be found calling the `getSize()` method. The
+   * `buffer` provided may be either a `nullptr`, in which case UCP will allocate a new
+   * memory region for it, or an already existing allocation, in which case UCP will only
+   * map it for RMA.
+   *
+   * @code{.cpp}
+   * // `context` is `std::shared_ptr<ucxx::Context>`
+   * // Allocate a 128-byte buffer with UCP.
+   * auto memoryHandle = context->createMemoryHandle(128, nullptr);
+   *
+   * // Equivalent to line above
+   * // auto memoryHandle = ucxx::createMemoryHandle(context, 128, nullptr);
+   *
+   * // Map an existing 128-byte buffer with UCP.
+   * size_t allocationSize = 128;
+   * auto buffer = new uint8_t[allocationSize];
+   * auto memoryHandleFromBuffer = context->createMemoryHandle(
+   *    allocationSize * sizeof(*buffer), reinterpret_cast<void*>(buffer)
+   * );
+   *
+   * // Equivalent to line above
+   * // auto memoryHandleFromBuffer = ucxx::createMemoryHandle(
+   * //    context, allocationSize * sizeof(*buffer), reinterpret_cast<void*>(buffer)
+   * // );
+   * @endcode
+   *
+   * @throws ucxx::Error if either `ucp_mem_map` or `ucp_mem_query` fail.
+   *
+   * @param[in] context parent context where to map memory.
+   * @param[in] size    the minimum size of the memory allocation
+   * @param[in] buffer  the pointer to an existing allocation or `nullptr` to allocate a
+   *                    new memory region.
```

Comment:
*suggestion*: Mention that if a buffer is provided, it is the user's responsibility to keep it alive until the memory handle is disposed of.